source/MRCuda/MRCudaMath.cuh

```diff
@@ -151,5 +167,167 @@ __device__ inline ClosestPointRes closestPointInTriangle( const float3& p, const
     return { { v, w }, a + ab * v + ac * w };
 }
 
+__device__ inline bool rayBoxIntersect( const Box3& box, const float3& rayOrigin, float& t0, float& t1, const IntersectionPrecomputes& prec )
+{
+    const int3& sign = prec.sign;
+
+    // compare and update x-dimension with t0-t1
+    t1 = min( ( box[sign.x].x - rayOrigin.x ) * prec.invDir.x, t1 );
+    t0 = max( ( box[1 - sign.x].x - rayOrigin.x ) * prec.invDir.x, t0 );
+
+    // compare and update y-dimension with t0-t1
+    t1 = min( ( box[sign.y].y - rayOrigin.y ) * prec.invDir.y, t1 );
+    t0 = max( ( box[1 - sign.y].y - rayOrigin.y ) * prec.invDir.y, t0 );
+
+    // compare and update z-dimension with t0-t1
+    t1 = min( ( box[sign.z].z - rayOrigin.z ) * prec.invDir.z, t1 );
+    t0 = max( ( box[1 - sign.z].z - rayOrigin.z ) * prec.invDir.z, t0 );
+    return t0 <= t1;
+}
+
+__device__ inline float rayTriangleIntersect(const float* oriA, const float* oriB, const float* oriC, const IntersectionPrecomputes& prec)
+{
+    const float Sx = prec.Sx;
+    const float Sy = prec.Sy;
+    const float Sz = prec.Sz;    
+
+    const float Ax = oriA[prec.idxX] - Sx * oriA[prec.maxDimIdxZ];
+    const float Ay = oriA[prec.idxY] - Sy * oriA[prec.maxDimIdxZ];
+    const float Bx = oriB[prec.idxX] - Sx * oriB[prec.maxDimIdxZ];
+    const float By = oriB[prec.idxY] - Sy * oriB[prec.maxDimIdxZ];
+    const float Cx = oriC[prec.idxX] - Sx * oriC[prec.maxDimIdxZ];
+    const float Cy = oriC[prec.idxY] - Sy * oriC[prec.maxDimIdxZ];
+
+    // due to fused multiply-add (FMA): (A*B-A*B) can be different from zero, so we need epsilon
+    const float eps = FLT_EPSILON * max( Ax, Bx, Cx, Ay, By, Cy );
+
+    float U = Cx * By - Cy * Bx;
+    float V = Ax * Cy - Ay * Cx;
+    float W = Bx * Ay - By * Ax;
+
+    if ( U < -eps || V < -eps || W < -eps )
+    {
+        if ( U > eps || V > eps || W > eps )
+        {
+            // U,V,W have clearly different signs, so the ray misses the triangle
+            return -FLT_MAX;
+        }
+    }
+
+    float det = U + V + W;
+    if ( det == 0 )
+        return -FLT_MAX;
+
+    const float Az = Sz * oriA[prec.maxDimIdxZ];
+    const float Bz = Sz * oriB[prec.maxDimIdxZ];
+    const float Cz = Sz * oriC[prec.maxDimIdxZ];
+    const float t = U * Az + V * Bz + W * Cz;
+    
+    return t / det;
+}
+
+__device__ inline int rayMeshIntersect( const Node3* nodes, const float3* meshPoints, const FaceToThreeVerts* faces, const float3& rayOrigin, float rayStart, float rayEnd, const IntersectionPrecomputes& prec )
```

Comment:
why have inmplementation in cuh file better make MRCudaRayCast.cuh and MRCudaRayCast.cu